tower/src/macros.rs

```diff
@@ -0,0 +1,24 @@
+macro_rules! opaque_future {
+    ($(#[$m:meta])* pub type $name:ident<$($param:ident),+> = $actual:ty;) => {
+        #[pin_project::pin_project]
+        $(#[$m])*
+        pub struct $name<$($param),+>(#[pin] pub(crate) $actual);
+
+        impl<$($param),+> std::fmt::Debug for $name<$($param),+> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_tuple(stringify!($name)).field(&format_args!("...")).finish()
+            }
+        }
+
+        impl<$($param),+> std::future::Future for $name<$($param),+>
+        where
+            $actual: std::future::Future,
+        {
+            type Output = <$actual as std::future::Future>::Output;
+            #[inline]
+            fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+                self.project().0.poll(cx)
+            }
+        }
+    }
+}
```

Comment:
The idea behind this thing is that we probably don't want to expose type aliases for `futures_util` futures in public APIs, for stability reasons. Since type aliases don't hide the aliased type, users could refer to these as the nested chain of `futures_util` combinator types. Thus, if we change the specific combinator futures, that's potentially a breaking change.

I'll wrap other exposed combinator futures in `opaque_future!` when this merges.